flink-core/src/main/java/org/apache/flink/api/common/state/StateDescriptor.java

```diff
@@ -82,8 +88,8 @@
 	/** The serializer for the type. May be eagerly initialized in the constructor,
 	 * or lazily once the {@link #initializeSerializerUnlessSet(ExecutionConfig)} method
 	 * is called. */
-	@Nullable
-	protected TypeSerializer<T> serializer;
+	@VisibleForTesting
+	protected final AtomicReference<TypeSerializer<T>> serializerAtomicReference = new AtomicReference<>();
```

Comment:
It's still a bit dangerous to make this `protected`, IMO. We can't guarantee subclasses would not mess around with this.

If we want to make this visible for testing, a separate package-private final method for getting the original wrapped serializer is better.

WDYT?

Reply:
I was a little bit struggling about reducing the visibility but I agree exposing this original serializer is dangerous. So I agree to reduce the visibility to private.

And since `StateDescriptor` is annotated as `PublicEvolving`, will add a release note about this on JIRA. Thanks.